cpu/stm32l1/periph/gpio.c

```diff
@@ -24,545 +26,196 @@
-            GPIO_6_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_7_EN
-        case GPIO_7:
-            GPIO_7_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_8_EN
-        case GPIO_8:
-            GPIO_8_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_9_EN
-        case GPIO_9:
-            GPIO_9_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_10_EN
-        case GPIO_10:
-            GPIO_10_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_11_EN
-        case GPIO_11:
-            GPIO_11_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_12_EN
-        case GPIO_12:
-            GPIO_12_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_13_EN
-        case GPIO_13:
-            GPIO_13_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_14_EN
-        case GPIO_14:
-            GPIO_14_EXTI_CFG();
-            break;
-#endif
-#ifdef GPIO_15_EN
-        case GPIO_15:
-            GPIO_15_EXTI_CFG();
-            break;
-#endif
+    else {
+        NVIC_EnableIRQ(EXTI15_10_IRQn);
     }
-
-    NVIC_EnableIRQ(gpio_irq_map[dev]);
-
-    /* set callback */
-    gpio_config[dev].cb = cb;
-    gpio_config[dev].arg = arg;
-
-    /* configure the event that triggers an interrupt */
+    /* configure the active edge(s) */
     switch (flank) {
         case GPIO_RISING:
-            EXTI->RTSR |= (1 << pin);
-            EXTI->FTSR &= ~(1 << pin);
+            EXTI->RTSR |= (1 << pin_num);
+            EXTI->FTSR &= ~(1 << pin_num);
             break;
         case GPIO_FALLING:
-            EXTI->RTSR &= ~(1 << pin);
-            EXTI->FTSR |= (1 << pin);
+            EXTI->RTSR &= ~(1 << pin_num);
+            EXTI->FTSR |= (1 << pin_num);
             break;
         case GPIO_BOTH:
-            EXTI->RTSR |= (1 << pin);
-            EXTI->FTSR |= (1 << pin);
+            EXTI->RTSR |= (1 << pin_num);
+            EXTI->FTSR |= (1 << pin_num);
             break;
     }
-
+    /* enable specific pin as exti sources */
+    SYSCFG->EXTICR[pin_num >> 2] &= ~(0xf << ((pin_num & 0x03) * 4));
+    SYSCFG->EXTICR[pin_num >> 2] |= (port_num << ((pin_num & 0x03) * 4));
     /* clear any pending requests */
-    EXTI->PR = (1 << pin);
-    /* unmask the pins interrupt channel */
-    EXTI->IMR |= (1 << pin);
-
+    EXTI->PR = (1 << pin_num);
+    /* enable interrupt for EXTI line */
+    EXTI->IMR |= (1 << pin_num);
     return 0;
 }
 
-void gpio_irq_enable(gpio_t dev)
+void gpio_init_af(gpio_t pin, gpio_af_t af)
 {
-    uint8_t pin;
+    GPIO_TypeDef *port = _port(pin);
+    uint32_t pin_num = _pin_num(pin);
 
-    if (dev >= GPIO_NUMOF) {
-        return;
-    }
-
-    pin = gpio_pin_map[dev];
-    EXTI->IMR |= (1 << pin);
+    /* set pin to AF mode */
+    port->MODER &= ~(3 << (2 * pin_num));
+    port->MODER |= (2 << (2 * pin_num));
+    /* set selected function */
+    port->AFR[(pin_num > 7) ? 1 : 0] &= ~(0xf << ((pin_num & 0x07) * 4));
+    port->AFR[(pin_num > 7) ? 1 : 0] |= (af << ((pin_num & 0x07) * 4));
 }
```

Comment:
I forgot the exact reason but IIRC @haukepetersen decided to enable the clock for the selected port here as well.